packages/core/src/test/shared/sam/utils.test.ts

```diff
@@ -148,4 +163,113 @@ describe('SAM utils', async function () {
+
+        it('returns path and version when SAM CLI is found and version is valid', async () => {
+            const expectedPath = '/usr/local/bin/sam'
+            const expectedVersion = '1.99.0'
+            getOrDetectSamCliStub.resolves({ path: expectedPath })
+            executeStub.resolves({ version: expectedVersion })
+
+            const result = await getSamCliPathAndVersion()
+
+            assert.strictEqual(result.path, expectedPath)
+            assert.strictEqual(result.parsedVersion?.version, expectedVersion)
+            assert(telemetryStub.calledOnceWith({ version: expectedVersion }))
+        })
+
+        it('throws MissingExecutable error when SAM CLI path is undefined', async () => {
+            getOrDetectSamCliStub.resolves({ path: undefined })
+
+            try {
+                await getSamCliPathAndVersion()
+                assert.fail('Should have thrown an error')
+            } catch (error) {
+                assert(error instanceof ToolkitError)
+                assert.strictEqual(error.code, 'MissingExecutable')
+                assert.strictEqual(error.message, 'SAM CLI could not be found')
+                assert(telemetryStub.notCalled)
+            }
+        })
+
+        it('throws VersionTooLow error when SAM CLI version is below 1.53.0', async () => {
+            const lowVersion = '1.52.0'
+            getOrDetectSamCliStub.resolves({ path: '/usr/local/bin/sam' })
+            executeStub.resolves({ version: lowVersion })
+
+            try {
+                await getSamCliPathAndVersion()
+                assert.fail('Should have thrown an error')
+            } catch (error) {
+                assert(error instanceof ToolkitError)
+                assert.strictEqual(error.code, 'VersionTooLow')
+                assert.strictEqual(error.message, 'SAM CLI version 1.53.0 or higher is required')
+                // records telemetry even when version is too low
+                assert(telemetryStub.calledOnceWith({ version: lowVersion }))
+            }
+        })
+    })
+
+    describe('get/update recent response', () => {
+        // Create stub for globals.context.workspaceState.get
+        const mementoRootKey = 'samcli.utils.key'
+        const nonExistingMementoRootKey = 'samcli.utils.no.key'
+        const identifier = 'us-east-1'
+        const key1 = 'stackName'
+        const key2 = 'bucketName'
+        const value1 = 'myStackName'
+        const value2 = 'myBucketName'
+
+        it('1. getRecentResponse should return undefined when mementoRootKey does not exist', async () => {
+            assert(!getRecentResponse(nonExistingMementoRootKey, identifier, key1))
+        })
+
+        it('2. updateRecentResponse should return the undefined when mementoRootKey does not exist', async () => {
+            try {
+                await updateRecentResponse(mementoRootKey, identifier, key1, value1)
+            } catch (err) {
+                assert.fail('Should have succeed')
```

Comment:
why we have a assert.fail here, from description it seems you want the test to fail if error happens.